File: app/src/main/java/com/tory/noname/activity/base/BaseActivity.java
package com.tory.noname.activity.base;

import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.design.widget.AppBarLayout;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;

import com.tory.noname.R;
import com.tory.noname.utils.StatusBarHelper;
import com.tory.noname.utils.Utilities;

public abstract class BaseActivity extends AppCompatActivity  {

    protected final String TAG = this.getClass().getSimpleName();

    protected Toolbar mToolbar;
    protected String mTitle;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        int layoutId = bindLayout();
        if(layoutId>0){
            setContentView(layoutId);
        }
        setThemeColor();
        initView();
        doBusiness();

    }

    protected void setThemeColor(){
        setStatusBar();
    }

    protected void initToolbar(){
        mToolbar = (Toolbar) findViewById(R.id.toolbar);
        if(mToolbar != null){
            setSupportActionBar(mToolbar);
        }
    }

    public void setToolbarTitle(String title) {
        if (mToolbar != null) {
            this.mTitle = title;
            //  mToolbar.setTitle(title);
            getSupportActionBar().setTitle(mTitle);
        }
    }

    public String getToolbarTitle(){
        return mTitle;
    }

    protected void setToolbarScrolled(boolean scrolled){
        if(mToolbar != null && mToolbar.getLayoutParams() instanceof AppBarLayout.LayoutParams){
            AppBarLayout.LayoutParams lp = (AppBarLayout.LayoutParams) mToolbar.getLayoutParams();
            if(scrolled){
                lp.setScrollFlags(AppBarLayout.LayoutParams.SCROLL_FLAG_SCROLL
                        |AppBarLayout.LayoutParams.SCROLL_FLAG_ENTER_ALWAYS);
            }else{
                lp.setScrollFlags(0);
            }

            mToolbar.setLayoutParams(lp);
        }
    }

    protected void setDisplayHomeAsUpEnabled(boolean enable){
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(enable);
            actionBar.setDisplayShowHomeEnabled(false);
        }
    }

    protected void setStatusBar(){
        if(Utilities.ATLEAST_LOLLIPOP) return;
        DrawerLayout drawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);
        int statusbarColor = ContextCompat.getColor(this,R.color.statusbarColor);
        if(drawerLayout == null){
            StatusBarHelper.setColor(this,statusbarColor);
        }else{
            StatusBarHelper.setColorForDrawerLayout(this,drawerLayout,statusbarColor);
        }
    }

    /**
     * 绑定渲染视图的布局文件
     *
     * @return 布局文件资源id
     */
    @LayoutRes
    public abstract int bindLayout();

    /**
     * 初始化控件
     */
    public abstract void initView();

    /**
     * 业务处理操作（onCreate方法中调用）
     *
     */
    public abstract void doBusiness();

    /**
     * Actionbar点击返回键关闭事件
     */

}

File: base/common/src/main/java/com/tory/library/base/BaseActivity.kt
package com.tory.library.base

import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.util.AttributeSet
import android.util.Log
import android.view.View
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.os.TraceCompat
import androidx.fragment.app.Fragment
import com.tory.library.R
import com.tory.library.utils.SystemBarUtils

abstract class BaseActivity : AppCompatActivity() {
    @JvmField
    protected val TAG = this.javaClass.simpleName

    var toolbar: Toolbar? = null
        protected set
    protected var mTitle: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        onCreateView(savedInstanceState)
        setThemeColor()
        initToolbar()
        initView(savedInstanceState)
        initData(savedInstanceState)
    }

    /**
     * 创建View
     */
    open fun onCreateView(savedInstanceState: Bundle?) {
        val layoutId = getLayoutId()
        if (layoutId > 0) {
            setContentView(layoutId)
        }
    }

    protected open fun setThemeColor() {
        SystemBarUtils.setStatusBarColor(this, Color.WHITE)
        SystemBarUtils.setStatusBarDarkMode(this, true)
        SystemBarUtils.setNavigationBarDarkMode(this, true)
    }
    protected open fun initToolbar() {
        toolbar = findViewById<Toolbar>(R.id.toolbar)
        if (toolbar != null) {
            setSupportActionBar(toolbar)
        }
        setDisplayHomeAsUpEnabled(true)
        setToolbarBackpress()
    }

    //监听toolbar左上角后退按钮
    fun setToolbarBackpress() {
        toolbar?.setNavigationOnClickListener { v: View? -> onBackPressed() }
    }

    //  mToolbar.setTitle(title);
    var toolbarTitle: String?
        get() = mTitle
        set(title) {
            if (toolbar != null) {
                mTitle = title
                //  mToolbar.setTitle(title);
                supportActionBar!!.title = mTitle
            }
        }

    protected fun setDisplayHomeAsUpEnabled(enable: Boolean) {
        val actionBar = supportActionBar
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(enable)
            actionBar.setDisplayShowHomeEnabled(false)
        }
    }

    fun showFragment(tag: String, show: Boolean = true, executeImmediately: Boolean = false) {
        TraceCompat.beginSection("showFragment - $tag")
        val fm = supportFragmentManager
        var fragment = fm.findFragmentByTag(tag)
        if (!show && fragment == null) {
            // Nothing to show, so bail early.
            return
        }
        val transaction = fm.beginTransaction()
        if (show) {
            if (fragment == null) {
                Log.d(TAG, "showFragment: fragment need create: $tag")
                fragment = createNewFragmentForTag(tag)
                transaction.add(getFragmentContainer(tag), fragment, tag)
            } else {
                Log.d(TAG, "showFragment: fragment is all ready created $tag")
                transaction.show(fragment)
            }
        } else {
            transaction.hide(fragment!!)
        }
        transaction.commitAllowingStateLoss()
        if (executeImmediately) {
            fm.executePendingTransactions()
        }
        TraceCompat.endSection()
    }

    private fun showOsFragment(tag: String, show: Boolean, executeImmediately: Boolean) {}
    fun getFragmentManagerByTag(tag: String?): Any {
        return supportFragmentManager
    }

    open fun getFragmentContainer(tag: String): Int {
        throw IllegalStateException("Unexpected fragmentContainer: $tag")
    }

    open fun createNewFragmentForTag(tag: String): Fragment {
        throw IllegalStateException("Unexpected fragment: $tag")
    }

    /**
     * 绑定渲染视图的布局文件
     *
     * @return 布局文件资源id
     */
    @LayoutRes
    open fun getLayoutId(): Int = 0

    /**
     * 初始化控件
     */
    abstract fun initView(savedInstanceState: Bundle?)

    /**
     * 业务处理操作（onCreate方法中调用）
     *
     */
    open fun initData(savedInstanceState: Bundle?) = Unit
}
